### ISSUE TITLE
A global switch to disable tensor broadcast for debugging purpose

### ISSUE THREAD

**Issue by Stonesjtu** [pytorch/pytorch#21847]
## 🚀 Feature

```python
torch.disable_broadcast()
# raise errors instead of doing broadcast underneath.

```

## Motivation

Broadcast is very convenient and efficient in many scenarios, but it also introduces extra careless bugs which are hard to locate.
It should be great if I can disable the broadcast mechanism (or force the Tensor size check) at debugging time.

**Comment by umanwizard** [comment]
I agree that this would be useful, but it might be non-trivial to implement, since not all operators use the same mechanism for doing broadcasting.

**Comment by ezyang** [comment]
Furthermore, a lot of internal code relies on broadcasting; it would all stop working if you could actually "disable" broadcasting.

Maybe what you actually want is just have some sort of logging mode, where it tells you about all the broadcasting that is taking place in a program. So I guess we could accept a PR like this, but you should talk to, e.g., @VitalyFedyunin about design first before attempting it.
